cli/azd/cmd/sets.go

```diff
@@ -0,0 +1,208 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/templates"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/azcli"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/git"
+	"github.com/google/wire"
+	"github.com/mattn/go-colorable"
+	"github.com/mattn/go-isatty"
+	"github.com/spf13/cobra"
+)
+
+func newWriter(cmd *cobra.Command) io.Writer {
+	writer := cmd.OutOrStdout()
+
+	if os.Getenv("NO_COLOR") != "" {
+		writer = colorable.NewNonColorable(writer)
+	}
+
+	// To support color on windows platforms which don't natively support rendering ANSI codes
+	// we use colorable.NewColorableStdout() which creates a stream that uses the Win32 APIs to
+	// change colors as it interprets the ANSI escape codes in the string it is writing.
+	if writer == os.Stdout {
+		writer = colorable.NewColorableStdout()
+	}
+
+	return writer
+}
+
+func newConsoleFromOptions(
+	rootOptions *internal.GlobalCommandOptions,
+	formatter output.Formatter,
+	writer io.Writer,
+	cmd *cobra.Command,
+) input.Console {
+	isTerminal := cmd.OutOrStdout() == os.Stdout &&
+		cmd.InOrStdin() == os.Stdin && isatty.IsTerminal(os.Stdin.Fd()) &&
+		isatty.IsTerminal(os.Stdout.Fd())
+
+	return input.NewConsole(!rootOptions.NoPrompt, isTerminal, writer, input.ConsoleHandles{
+		Stdin:  cmd.InOrStdin(),
+		Stdout: cmd.OutOrStdout(),
+		Stderr: cmd.ErrOrStderr(),
+	}, formatter)
+}
+
+func newCommandRunnerFromConsole(console input.Console) exec.CommandRunner {
+	return exec.NewCommandRunner(
+		console.Handles().Stdin,
+		console.Handles().Stdout,
+		console.Handles().Stderr,
+	)
+}
+
+func newAzCliFromOptions(rootOptions *internal.GlobalCommandOptions, cmdRun exec.CommandRunner) azcli.AzCli {
+	return azcli.NewAzCli(azcli.NewAzCliArgs{
+		EnableDebug:     rootOptions.EnableDebugLogging,
+		EnableTelemetry: rootOptions.EnableTelemetry,
+		CommandRunner:   cmdRun,
+		HttpClient:      nil,
+	})
+}
+
+func newAzdContext() (*azdcontext.AzdContext, error) {
+	azdCtx, err := azdcontext.NewAzdContext()
+	if err != nil {
+		return nil, fmt.Errorf("creating context: %w", err)
+	}
+
+	return azdCtx, nil
+}
+
+var FormattedConsoleSet = wire.NewSet(
+	output.GetCommandFormatter,
+	newWriter,
+	newConsoleFromOptions,
+)
+
+var CommonSet = wire.NewSet(
+	newAzdContext,
+	FormattedConsoleSet,
+	newCommandRunnerFromConsole,
+)
+
+var InitCmdSet = wire.NewSet(
+	CommonSet,
+	newAzCliFromOptions,
+	git.NewGitCliFromRunner,
+	newInitAction,
+	wire.Bind(new(actions.Action), new(*initAction)))
```

Comment:
Any reason to create & expose sets that are only ever going to be used in a single injector?  From my research sets become more powerful when then bundle commonly used components.  For example, the `CommonSet` makes a ton of sense since it exposes some core types that are used in many areas.

Also - instead of having a single wire.go, would it be better to have a wire.go file for each package? And then expose sets from each package that then be composed at higher levels within the app?  This would allow us to keep a better separation of concerns across the board.

Reply:
I have a sample of this @ https://github.com/wbreza/wire-sample/tree/main/src/pkg/provisioning

Reply:
> Any reason to create & expose sets that are only ever going to be used in a single injector? From my research sets become more powerful when then bundle commonly used components. For example, the CommonSet makes a ton of sense since it exposes some core types that are used in many areas.

Do you mean lowercase "initCmdSet"? instead of "InitCmdSet"? You'll need a `Set` eventually since that is what `wire.Build` accepts... In your example, `ProviderSet` is in `wire.go`. And the reason to not put in `wire.go` is that you don't get any intellisense feedback since `wire.go` is not compiled without `wireinject` flag being present.

> Also - instead of having a single wire.go, would it be better to have a wire.go file for each package? And then expose sets from each package that then be composed at higher levels within the app? This would allow us to keep a better separation of concerns across the board.

Would love to understand this better. The idea behind dependency injection is that libraries don't control their construction, and the app in `main.go` or similar does the construction.

Reply:
> Also - instead of having a single wire.go, would it be better to have a wire.go file for each package? And then expose sets from each package that then be composed at higher levels within the app? This would allow us to keep a better separation of concerns across the board.

OK. Looked into this more. The problem is that in that world, each package is coupled to other packages, so all packages are strictly "azd packages". Whereas here, each package can abstractly depend on an interface, and azd in main.go can wire up the actual package implementation.

The other problem is that you can't have the same provider registered more than once, so if you had:
`AzCliProviderSet = { CommonSet, NewAzCli }`

and separately:
`FrameworkServiceProviderSet = { CommonSet, NewFrameworkService }`

This would fail to generate:
`InfraCreateSet = { AzCliProviderSet, FrameworkServiceProviderSet }`

since `CommonSet` is registered twice.

Reply:
Makes sense - how the object are instantiated is more of an application concern and all the packages should be concerned with is that they require a set of dependencies and don't need to worry about the concern of their being IoC/DI involved at all.

Reply:
> Do you mean lowercase "initCmdSet" instead of "InitCmdSet"? You'll need a Set eventually since that is what wire.Build accepts... In your example, ProviderSet is in wire.go. And the reason to not put in wire.go is that you don't get any intellisense feedback since wire.go is not compiled without wireinject flag being present.

Looks like `wire.Build(...)` can take any list of injectory commands and does not need to be a set as long as the first param is the object you are constructing.

Reply:
> Looks like wire.Build(...) can take any list of injectory commands and does not need to be a set as long as the first param is the object you are constructing.

Yeah, you can do it inline but you still do lose intellisense once it's inside `wire.go`. I do find that a little annoying. I wish we only had 1 file to specify injection and not two.